### PR TITLE
[JENKINS-63833] Make the groovy script source customizable

### PR DESCRIPTION
**Pull request by PierreBtz**
~WIP, I need to test some more use cases and would like to add some tests... Opening for early discussion.~

See [JENKINS-63833](https://issues.jenkins-ci.org/browse/JENKINS-63833).

### Proposed changelog entries

* Developer: Make the root source paths of GroovyHookScript customizable with `jenkins.util.groovy.GroovyHookScript.ROOT_PATH`

### Proposed upgrade guidelines

N/A

### Submitter checklist

- [x] (If applicable) Jira issue is well described
- [x] Changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
  * Fill-in the `Proposed changelog entries` section only if there are breaking changes or other changes which may require extra steps from users during the upgrade
- [ ] Appropriate autotests or explanation to why this change has no tests
- ~[ ] For dependency updates: links to external changelogs and, if possible, full diffs~

### Desired reviewers

@oleg-nenashev 

### Maintainer checklist

Before the changes are marked as `ready-for-merge`: 

- [ ] There are at least 2 approvals for the pull request and no outstanding requests for change
- [ ] Conversations in the pull request are over OR it is explicit that a reviewer does not block the change
- [ ] Changelog entries in the PR title and/or `Proposed changelog entries` are correct
- [ ] Proper changelog labels are set so that the changelog can be generated automatically
- [ ] If the change needs additional upgrade steps from users, `upgrade-guide-needed` label is set and there is a `Proposed upgrade guidelines` section in the PR title. ([example](https://github.com/jenkinsci/jenkins/pull/4387))
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins-ci.org/issues/?filter=12146)).

### CODE REVIEW

File: core/src/main/java/hudson/init/impl/GroovyInitScript.java [hunk 1]
```diff
@@ -37,8 +39,17 @@
  * @author Kohsuke Kawaguchi
  */
 public class GroovyInitScript {
+    private static final String PATH = System.getProperty(GroovyInitScript.class.getName() + ".PATH");
+    private static final String DIRECTORY_PATH = System.getProperty(GroovyInitScript.class.getName() + ".DIRECTORY_PATH");
```
Comment (daniel-beck) [inline core/src/main/java/hudson/init/impl/GroovyInitScript.java:43]: Use `SystemProperties` for this.

**Review by daniel-beck — COMMENTED** [review]
This needs a more clearly spelled out justification.

**Comment by PierreBtz** [comment]
> This needs a more clearly spelled out justification.

Currently the only way to provide a hook script (init in my case, but might get more general in the future) or a hook script directory is through the `JENKINS_HOME`. In the context of the Jenkinsfile runner, this is an issue as there is not necessarily a `JENKINS_HOME` available. The current approach (developed in https://github.com/jenkinsci/jenkinsfile-runner/pull/370) is to make a copy of the provided hooks prior to execution which is not really good performance wise. Another approach discussed in the aforementioned PR ([see comment](https://github.com/jenkinsci/jenkinsfile-runner/pull/370#issuecomment-702345383)) would be to use symlinks but this might cause problems (eg Windows?). This approach has the merit of keeping thing simple on the Jenkinsfile runner side as there is no more hacks to perform.

**Review by oleg-nenashev — CHANGES_REQUESTED** [review]
Groovy Hooks represent an extensible engine in Jenkins. Right now the customization is supported only for Init Hooks, and it looks a bit odd. It would be great to have equivalend support for all types of hooks, even if it needs additional directory level (hooks are in `MY_CUSTOM_DIR/init.groovy.d/myHook.groovy`)

**Comment by MRamonLeon** [comment]
Added the WIP label per the description

**Comment by PierreBtz** [comment]
@oleg-nenashev except for init script, I only found hooks used for boot-failure scripts: https://github.com/search?q=org%3Ajenkinsci+GroovyHookScript&type=code. Is there any other entry point I'm missing appart for potential private plugin and other groovy scripts calling this?

**Comment by PierreBtz** [comment]
@oleg-nenashev I implemented your proposal, it's easier and more generic :)
I currently don't check if the provided directory exist so there is no log in case of a non existing directory, probably not ideal.

File: core/src/main/java/jenkins/util/groovy/GroovyHookScript.java [hunk 2]
```diff
@@ -38,10 +39,11 @@
  * @author Kohsuke Kawaguchi
  */
 public class GroovyHookScript {
+    private static final String ROOT_PATH= SystemProperties.getString(GroovyHookScript.class.getName() + ".ROOT_PATH");
```
Comment (daniel-beck) [inline core/src/main/java/jenkins/util/groovy/GroovyHookScript.java:42]:
```suggestion
    private static final String ROOT_PATH = SystemProperties.getString(GroovyHookScript.class.getName() + ".ROOT_PATH");
```

File: core/src/main/java/jenkins/util/groovy/GroovyHookScript.java [hunk 3]
```diff
@@ -53,10 +55,10 @@ private GroovyHookScript(String hook, Jenkins j) {
         this(hook, j.servletContext, j.getRootDir(), j.getPluginManager().uberClassLoader);
     }
 
-    public GroovyHookScript(String hook, @NonNull ServletContext servletContext, @NonNull File home, @NonNull ClassLoader loader) {
+    public GroovyHookScript(String hook, @NonNull ServletContext servletContext, @NonNull File jenkinsHome, @NonNull ClassLoader loader) {
         this.hook = hook;
         this.servletContext = servletContext;
-        this.home = home;
+        this.rootDir = ROOT_PATH != null ? new File(ROOT_PATH): jenkinsHome;
```
Comment (daniel-beck) [inline core/src/main/java/jenkins/util/groovy/GroovyHookScript.java:61]:
```suggestion
        this.rootDir = ROOT_PATH != null ? new File(ROOT_PATH) : jenkinsHome;
```

**Review by daniel-beck — APPROVED** [review]
Looks reasonable (even if the use case doesn't seem well explained to me).

Should have a corresponding jenkins.io PR adding documentation to https://github.com/jenkins-infra/jenkins.io/blob/master/content/doc/book/managing/system-properties.adoc

**Comment by PierreBtz** [comment]
@oleg-nenashev coming back to this, I think it's ready. I wanted to add tests but I don't see an easy way to test this unitary...feature wasn't tested up until now either...

**Review by oleg-nenashev — APPROVED** [review]
Sorry that I missed this pull request. This change is totally valid, and it achieves the desired behavior. Thanks a lot @PierreBtz , and sorry for such a long review delay

We may merge it in 24 hours if there is no negative feedback. Please see the [merge process documentation](https://github.com/jenkinsci/jenkins/blob/master/docs/MAINTAINERS.adoc#merge-process) for more information about the merge process

**Comment by daniel-beck** [comment]
Is there no downstream PR demonstrating how this is used? 😕

**Comment by PierreBtz** [comment]
@daniel-beck I added a doc in https://github.com/jenkins-infra/jenkins.io/pull/3992, I plan to do a PR for the JFR in the future as this was the primary driver for this PR.

**Comment by daniel-beck** [comment]
@PierreBtz Thanks for the documentation update! I meant the demonstration how this would be used in JFR.

**Comment by oleg-nenashev** [comment]
@daniel-beck No reference implementation per se. There is https://github.com/jenkinsci/jenkinsfile-runner/pull/370 , and in this PR copying files would be replaced by setting a flag. I hope it is enough to justify the PR

**Comment by oleg-nenashev** [comment]
I will go ahead and merge it. This is a relevant feature for Jenkins packaging use-ases

**Pull request merged**